### PR TITLE
[qunat][pt2e] Support allow_implicit_sharing flag

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113115
* __->__ #112929

Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 1]
```diff
@@ -1060,6 +1060,83 @@ def validate(self, model: torch.fx.GraphModule) -> None:
...
+        """This tests the allow_transitive_sharing flag of QuantizationAnnotation, that is
+        if a node is configured with allow_implicit_sharing=False, we will not have implicit sharing
+        for node and (node, consumer) even they refer to the same Tensor
+
+        x1 -> add1 -----> add3
+        x2 -/              /
+               x3 -> add2 /
+               x4 -/
+
+        all add has shared input and output, and second input is using shared quantization spec pointint
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_quantize_pt2e.py:1073]:
```suggestion
        all add has shared input and output, and second input is using shared quantization spec pointing
```

File: test/quantization/pt2e/test_quantize_pt2e.py [hunk 2]
```diff
@@ -1060,6 +1060,83 @@ def validate(self, model: torch.fx.GraphModule) -> None:
...
+        to first input, but we set allow_implicit_sharing to all add nodes so input and output of add1,
```
Comment (kimishpatel) [inline test/quantization/pt2e/test_quantize_pt2e.py:1074]: you mean set allow_implicit_sharing to false?

**Review by kimishpatel — APPROVED** [review]
Left some comments. looks good.

File: test/quantization/pt2e/test_quantize_pt2e.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/pt2e/test_quantize_pt2e.py:1074]: yeah, will fix

**Pull request closed without merging**